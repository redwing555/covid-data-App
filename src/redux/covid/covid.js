const FETCH_COVID_DATA = 'covid-data-app/covid/FETCH_COVID_DATA';
const FILTER_DATA = 'covid-data-app/covid/FETCH_DATA';

export const fetchCovidData = (payload) => ({
  type: FETCH_COVID_DATA,
  payload,
});

export const fetchFromApi = () => async (dispatch) => {
  const today = new Date().toISOString().slice(0, 10);
  const response = await fetch(`https://api.covid19tracking.narrativa.com/api/${today}`);
  const data = await response.json();
  dispatch(fetchCovidData(data));
};

export const filterData = (payload) => ({
  type: FILTER_DATA,
  payload,
});

const covidReducer = (state = [], action) => {
  switch (action.type) {
    case FETCH_COVID_DATA:
      return [...state, action.payload.dates];

    case FILTER_DATA:
      return action.payload;
    default:
      return state;
  }
};

export { covidReducer };
