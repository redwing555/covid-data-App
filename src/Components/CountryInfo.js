import React, { useEffect, useState } from 'react';
import {
  useHistory, Route, Switch, useRouteMatch,
} from 'react-router-dom';
import PropTypes from 'prop-types';
import { useDispatch, useSelector } from 'react-redux';
import LineChart from './Plot';
import { dataLoading, loadHistoryThunk } from '../redux/covid/covid';
import '../styles/countryInfo.css';

const CountryInfo = ({ current, image }) => {
  const dispatch = useDispatch();
  const historyData = useSelector((state) => state.covidReducer.currentHistory);

  const loading = useSelector((state) => state.covidReducer.loading);
  const country = useSelector((state) => state.covidReducer.countries[current]);
  const total = country.today_confirmed;

  const rows = [
    {
      id: 1,
      text: 'Confirmed daily cases',
      value: country.today_new_confirmed,
      url: 'newDailyCases',
      category: 'today_new_confirmed',
    },
    {
      id: 2,
      text: 'Confirmed daily deaths',
      value: country.today_new_deaths,
      url: 'newDailyDeaths',
      category: 'today_new_deaths',
    },
    {
      id: 3,
      text: 'Confirmed daily Recovered',
      value: country.today_new_recovered,
      url: 'newDailyRecovered',
      category: 'today_new_recovered',
    },
    {
      id: 4,
      text: 'Confirmed number of Cases',
      value: country.today_confirmed,
      url: 'confirmedCases',
      category: 'today_confirmed',
    },
    {
      id: 5,
      text: 'Confirmed number of Death',
      value: country.today_deaths,
      url: 'confirmedDeaths',
      category: 'today_deaths',
    },
    {
      id: 6,
      text: 'Confirmed recoveries',
      value: country.today_recovered,
      url: 'confirmedRecoveries',
      category: 'today_recovered',
    },

  ];

  const [plotData, setPlotData] = useState([]);
  const history = useHistory();
  const { path, url } = useRouteMatch();

  const handleRoute = (route) => {
    history.push(route);
  };

  const handlePlotData = (category) => {
    const data = Object.keys(historyData).map((date) => {
      const tmp = historyData[date].countries[current][category];
      return tmp;
    });
    setPlotData(data);
  };

  useEffect(() => {
    dispatch(dataLoading());
    dispatch(loadHistoryThunk(current));
  }, [dispatch]);

  return (
    <div className="info-container">
      <div className="header-wrapper">
        <div className="world-header">
          <img className="country-map-header" src={image} alt="world map" />

        </div>
        <div className="country-header-info">
          <span className="world-cases">{current}</span>
          <span>
            {' '}
            {total && Intl.NumberFormat('de-DE').format(total)}
            {' '}
            Cases

          </span>

        </div>

      </div>

      <Switch>
        <Route exact path={path}>
          <div className="data-of-wrapper">
            <span className="data-of">
              {' '}
              Stats of
              {' '}
              {country.date}
              {' '}
              for
              {' '}
              {current}
              {' '}
            </span>

          </div>
          <ul className="table-infos">
            <li className="table-header">
              <div className="date">Date</div>
              <div className="date-value">{country.date}</div>
            </li>
            {loading && <span>Getting data for you ... </span>}
            {!loading && rows.map((row) => (

              <li
                className="country-details"
                key={row.id}
                onClick={() => {
                  handleRoute(`${url}/${row.url}`);
                  handlePlotData(row.category);
                }}
                aria-hidden="true"
              >

                <div className="metric">{row.text}</div>
                <div className="metric-value">{row.value}</div>
              </li>
            ))}
          </ul>

        </Route>
        <Route plotData={plotData} exact path={`${path}/newDailyCases`}>
          <LineChart plotData={plotData} country={current} title="new Cases" />
        </Route>
        <Route plotData={plotData} exact path={`${path}/newDailyDeaths`}>
          <LineChart plotData={plotData} country={current} title="new Deaths" />
        </Route>
        <Route exact path={`${path}/newDailyRecovered`}>
          <LineChart plotData={plotData} country={current} title="new Recoveries" />
        </Route>
        <Route exact path={`${path}/confirmedCases`}>
          <LineChart plotData={plotData} country={current} title="today new Cases" />
        </Route>
        <Route exact path={`${path}/confirmedDeaths`}>
          <LineChart plotData={plotData} country={current} title="today new Deaths" />
        </Route>
        <Route exact path={`${path}/confirmedRecoveries`}>
          <LineChart plotData={plotData} country={current} title="today new Recoveries" />
        </Route>
      </Switch>

    </div>
  );
};

CountryInfo.defaultProps = {
  current: 'No Country',
};

CountryInfo.propTypes = {
  current: PropTypes.string,
  image: PropTypes.string.isRequired,

};

export default CountryInfo;
