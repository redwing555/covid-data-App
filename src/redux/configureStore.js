import { createStore, combineReducers, applyMiddleware } from 'redux';
import logger from 'redux-logger';
import thunk from 'redux-thunk';
import { covidReducer } from './covid/covid';
import { flagReducer } from './flags/flags';

const reducer = combineReducers({
  covidReducer,
  flagReducer,
});

const middleware = [logger, thunk];

const store = createStore(
  reducer,
  applyMiddleware(...middleware),
);

export default store;
