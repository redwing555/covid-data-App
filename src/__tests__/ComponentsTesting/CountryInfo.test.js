import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { BrowserRouter as Router, Route } from 'react-router-dom';
import logger from 'redux-logger';
import thunk from 'redux-thunk';
import { createStore, combineReducers, applyMiddleware } from 'redux';
import CountryInfo from '../../Components/CountryInfo';
import { covidReducer } from '../../redux/covid/covid';

const reducer = combineReducers({
  covidReducer,

});

const store = createStore(
  reducer,
  applyMiddleware(logger, thunk),
);
// import store from '../redux/configureStore';

describe('Country information', () => {
  it('renders', () => {
    const tree = renderer
      .create(

        <Provider store={store}>
          <Router>
            <Route path="/country">
              <CountryInfo
                current="Morocco"
                image="https://raw.githubusercontent.com/djaiss/mapsicon/33ba28808f8d32b5bae0ffada9cadd07073852e1/all/Morocco/vector.svg"
              />
            </Route>
          </Router>
        </Provider>

        ,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
