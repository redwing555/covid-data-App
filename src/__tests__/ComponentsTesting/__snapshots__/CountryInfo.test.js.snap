// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Country information renders 1`] = `null`;
